### ISSUE TITLE
Records with mismatched countries

### ISSUE THREAD

**Issue by trescube** [whosonfirst/whosonfirst-data#211]
Attached is a file containing WOF IDs that have an `iso2` property that doesn't match the country in the hierarchy.  The file format has 3 fields:

1.  WOF ID
2.  `iso2` property value
3.  hierarchy-referenced country `iso2` property

An example entry is:

`85632199,BL,FR`

This should be interpreted to mean that WOF ID `85632199` has an `iso2` property value of `BL` but it's hierarchy references country `FR`.  

Please note that WOF ID `85681265` has literal `-1` as it's `iso2` property.  

[mismatched_countries.txt](https://github.com/whosonfirst/whosonfirst-data/files/133646/mismatched_countries.txt)

**Comment by nvkelso** [comment]
@trescube Thanks for creating this list! Some of these are fixes over Quattroshapes (which hand coded some country codes incorrectly), and some of these are errors probably in how the PIP was setup using bad math "centroids" that weren't actually in the feature, and some are just funky. 

Can you confirm where the (2) iso2 property value is coming from versus the (3) hierarchy-referenced country iso2 property? 

For the example, `85632199`, Saint Barthelemy, a dependency, WOF gives:

```
iso
country	BL
parent	FR
```

Which is probably an unfortunate naming, as `country` I think is meaning `admin-0` inclusive of country and dependency, and `parent` here refers to `empire`. @thisisaaronland to confirm.

Similarly:

```
wof
country	BL
```

Looks like the "empire" of France hasn't been fully modeled yet, though?

* https://whosonfirst.mapzen.com/spelunker/id/136253037/

The -1 literal is valid WOF'ism for weird stuff:

* https://whosonfirst.mapzen.com/spelunker/id/85681265/

**Comment by trescube** [comment]
Correct, the iso2 property I'm referring to is coming from `iso:country`.  We actually use iso3 for country code in our WOF importer but since that data isn't available (at least not easily) we use an external library to convert iso2->iso3.  So far it's just been easier to use the `iso2:country` field in the record but I suppose we could use that field from the country hierarchy element.  Which approach is advisable?

**Comment by trescube** [comment]
This is especially problematic for the 5,785 entries like:

`101878271,UK,UA`

Where the `iso2` value is UK whereas the correct value is UA.

**Comment by thisisaaronland** [comment]
Interesting. This appears to have been part of a batch update in September, specifically this one where 250K records were updated all at once.

https://github.com/whosonfirst/whosonfirst-data/commit/2f1666ea58545ee23d5def145b0a90a342be9638#diff-908a336350ffb813044109ffae81bf01

I will run some sanity checking with your list and the current PIP server. I am happy to defer to your list for `iso:country` but `wof:country` will deviate from the ISO country code for the reasons that @nvkelso mentioned above.

As far as the Baykonur Cosmodrome, I can update that by hand but it will be `XZ` per this document and is the country code equivalent of `-2` (which is what that records parent ID should be...)

https://github.com/whosonfirst/whosonfirst-placetypes#iso-country-codes

**Comment by thisisaaronland** [comment]
Of the 21704 records, all but 2087 are localities.

Of the 19617 (localities) there are 1082 where the `(pip) ISO:country != (parent) ISO:country == (parent) WOF:country` leaving 18535 records that can be updated with confidence.

So I will do that shortly, after dumping the (2087) not-localities and (1082) weird-localities in to their own CSV files for future processing.

**Comment by thisisaaronland** [comment]
Appended with `.txt` because GitHub's uploader is afraid of things ending in `.csv` ... because computers.

[Uploading wof-211-not-loc.csv.txt…]()
[Uploading wof-211-weird-loc.csv.txt…]()

**Comment by thisisaaronland** [comment]
Bucketed by placetype:

```
python ./pt.py ./wof-211-not-loc-xx.csv 
dependency: 1
region: 10
disputed: 38
neighbourhood: 26
```

These are things where the bulk are WOF `iso:country=XX` and the parent/hierarchy will be a specific country. This accounts for the fact that in most of the cases the placetype is disputed so there is no single ISO country code.

The neighbourhoods are all in India. As in WOF says `IN` and the initial bug report says `XX` @trescube – how did you arrive at a designation of `XX` ?

These are the remaining not-localities with `XX` codes filtered out. The large number of neighbourhoods suggests that there might be an easy fix for these?

```
python ./pt.py ./wof-211-not-loc-2.csv 
localadmin: 61
region: 144
county: 108
dependency: 37
macroregion: 18
neighbourhood: 1644
```

**Comment by stepps00** [comment]
This will be a next up, as we're now bucketing records by country code.
